kong/tracing/request_id.lua

```diff
@@ -0,0 +1,144 @@
+local to_hex = require "resty.string".to_hex
+local base = require "resty.core.base"
+
+local get_request = base.get_request
+
+
+-- Request ID types
+-- higher value means higher priority
+-- CORR  = correlation ID
+-- TRACE = trace ID
+-- INIT  = the initial value
+local TYPES = {
+  INIT = {
+    name = "initial",
+    priority = 1,
+  },
+  TRACE = {
+    name = "trace",
+    priority = 2,
+  },
+  CORR = {
+    name = "correlation",
+    priority = 3,
+  },
+}
+
+local NGX_VAR_PHASES = {
+  set           = true,
+  rewrite       = true,
+  access        = true,
+  content       = true,
+  header_filter = true,
+  body_filter   = true,
+  log           = true,
+  balancer      = true,
+}
+
+
+local function get_ctx_request_id()
+  return ngx.ctx.request_id
+end
+
+
+local function set_ctx_request_id(id, type)
+  ngx.ctx.request_id = {
+    id = id,
+    type = type,
+  }
+end
+
+
+--- Function `should_overwrite_id` determines whether the new request id
+-- should overwrite the current one.
+--
+-- This function uses the provided (new) request id type to compare it with the
+-- current (old) request id's and determine whether the new has more priority
+-- than the old and should overwrite it.
+--
+-- @function should_overwrite_id
+-- @param new_type The new request id's type
+-- @param old_type The old request id's type
+-- @return boolean whether the new request id should overwrite the current one
+local function should_overwrite_id(new_type, old_type)
+  if new_type.priority < old_type.priority then
+    kong.log.info("request_id of type: ", new_type.name,
+                  "is less prioritary than current type: ", old_type.name)
+    return false
+  end
+
+  return true
+end
+
+
+local function get()
+  if not get_request() then
+    return nil, "no request found"
+  end
+
+  local rid = get_ctx_request_id()
+  rid = rid and rid.id
+
+  if not rid then
+    -- first access to the request id for this request:
+    -- try to initialize with the value of $kong_request_id
+    local ok
+    ok, rid = pcall(function() return ngx.var.kong_request_id end)
```

Comment:
I think that `ngx.var.xx` will always success, what is the purpose of `pcall`?

Reply:
this used to be accessed in phases that would not allow accessing ngx.var, but that was an older implementation, the pcall can now be removed, thanks.